### PR TITLE
Trying to enable IMS services on (almost) all mobile service providers

### PR DESCRIPTION
**Pull request by antoniu200**
Hi,

I am currently trying to enable the aformentioned features in this image. However, I cannot yet test by myself, since simply pulling the source code proved to be so much more problematic than I anticipated, I ended up wasting 2 days and the source code still isn't pulled correctly & completely.

Anyway, DIGI clearly told me on the phone that this phone is not compatible with their network, but their network also doesn't check what type of device is trying to connect to VoLTE / VoWiFi, so if I know how to make it work (I don't, but I can try), it should work.

As for Orange, they have provided Sony with a modem MBN and VoLTE works and is enabled. However, this image does not automatically switch to their MBN and VoLTE only works on that modem MBN (no `modem.conf` present linking to MBN, it seems + `OpenCustomizationSelector` not configured). Also, VoWiFi is not enabled in CarrierConfig overlay.

I made some modifications to the `vendor.xml` in CarrierConfig.

The reason for this pull request is that if, by any chance, you manage to build a test image sooner than I can, it would be amazing if you could include these modifications in it and send me a link to the .zip (or whatever the end result of the compilation is - never compiled Android before), so I can test whether my modifications helped or not.

### CODE REVIEW

**Comment by the-r3dacted** [comment]
You'll need to do more than that for DIGI. You'll need to set or make OpenCustomizationSelector automatically set the ir51_ir92 mbn for VoLTE. Try testing on stock ROM using a method similar to below.
https://xdaforums.com/t/how-to-get-t-mobile-volte-working-on-the-xa2.4647892/

**Comment by antoniu200** [comment]
Thanks for the reply!

> set the ir51_ir92 mbn for VoLTE

Are you sure this is required? Afaik, `the ir51_ir92` modem configuration has that dreaded modem bug (I get no signal on DIGI using it, so I know no further), while the stock modem does not.

> Try testing on stock ROM using a method similar to below.

As for that method, I had already tried that a while before doing the above changes in code - it did not work. What is curious is that neither did Orange's VoLTE (which is officially supported on this phone) without adding some extra lines to system.prop to force enable VoLTE and VoWiFi toggles (which is what that .zip does).
OpenCustomizationSelector also seems to fail to automatically switch to `amss_fsg_lilac_orange_romania_ims_tar.mbn` when inserting an Orange Romania SIM, which is the correct modem. It rather switches to the normal `amss_fsg_lilac_tar.mbn`.

Unfortunately, I never got to compiling this version of LineageOS, since it has a bug that annoys me to death: the Navbar is overlapping the Status bar.
I have no idea how to fix that, I [asked on XDA - nobody replied](https://xdaforums.com/t/status-bar-overlaps-navigation-bar-in-a-build-of-lineage-os-17-1-how-can-it-be-fixed.4644815/), I even found [this old thread](https://xdaforums.com/t/status-bar-overlaps-navigation-bar.4283983), but nobody said anything about it. However, Google issued an update and that ended up fixing it. Nobody knows, afaik, what was actually fixed.

So, what I tried to do was compile a newer version of AOSP than the one used by the base of LineageOS 17.1 (r47 vs. r41), but I am getting a bunch of errors. I do not remember them exactly.

If you could help me get through the compiling process, I would be able to test and show you the exact errors I am getting. Otherwise, I will have to ask around on my own time, which is very scarce these 3 coming weeks.

**Comment by the-r3dacted** [comment]
> > set the ir51_ir92 mbn for VoLTE
> 
> Are you sure this is required? Afaik, `the ir51_ir92` modem configuration has that dreaded modem bug (I get no signal on DIGI using it, so I know no further), while the stock modem does not.

That's not the modem bug. I'd check Network Signal Guru to see what it says. The modem bug is seemingly a US only thing where the modem keeps restarting on the T-Mobile network. I'd be curious to do more testing with 8 based custom ROMs, but I haven't gotten around to it. (My big issue with them is the lowest brightness is too bright for me to use at night)
 
> > Try testing on stock ROM using a method similar to below.
> 
> As for that method, I had already tried that a while before doing the above changes in code - it did not work. What is curious is that neither did Orange's VoLTE (which is officially supported on this phone) without adding some extra lines to system.prop to force enable VoLTE and VoWiFi toggles (which is what that .zip does).

Yeah, I guess it didn't apply those lines properly, so you have to.

> Unfortunately, I never got to compiling this version of LineageOS, since it has a bug that annoys me to death: the Navbar is overlapping the Status bar. I have no idea how to fix that,
> 
> If you could help me get through the compiling process, I would be able to test and show you the exact errors I am getting.

Weird bug. I've never been able to build this ROM. The extract files process always fails. I've been able to build 17.1 for the XZ2C using a manifest that just downloads the proprietary blobs instead of extracting them. Didn't see the statusbar issue there.

Edit: I should look at things before I mention them. I used the phone in landscape for a bit and didn't remember that being an issue. I checked it, and it does have the issue. Maybe we could find the commit that fixed it in AOSP and make it into a .patch file for LineageOS.

**Comment by antoniu200** [comment]
Okay, I am finally back.

I have tried the `amss_fsg_lilac_ir_51_ir_92_ims_tar.mbn` again. Maybe my mind was playing tricks on me before, because it now worked on both DIGI and Orange.
Furthermore, on DIGI it finally registered the phone in IMS. However, although it lists voth VoWiFi and VoLTE as being available in `*#*#4636#*#*`, it does not make calls using them. It seems like it needs further configuration (possibly OpenCustomizationSelector).

On Orange, with their proprietary `amss_fsg_lilac_orange_romania_ims_tar.mbn`, only VoLTE works. With `amss_fsg_lilac_ir_51_ir_92_ims_tar.mbn`, both VoLTE and VoWiFi work without any further modifications.

Question: **Do you have any clue what app is responbile for switching the Modem Customization?** Currently, it does not switch correctly, it always tries to switch to `amss_fsg_lilac_tar.mbn`, even though a different one is marked as correct for the ISP.

**Comment by the-r3dacted** [comment]
It's OpenCustomizationSelector. I would think manually picking an mbn would just set that one but guess I'm wrong. You could rename the ir92 or whatever mbn to replace stock and have it work that way. I'm not really sure where the OpenCustomizationSelector app is stored in this ROM. I tried looking for it in various repos, but I didn't find it in any.

**Comment by antoniu200** [comment]
Alright, thanks! Manually picking does work, but OpenCustomizationSelector still resets the custom modem to default if the SIM is changed, which is very frustrating, especially since for Orange it's the wrong modem and, thus, requires 3 reboots: one for changing the SIM, one for wrong config, one for correct config.

I found the app, it's in 2 locations, one for `yoshino-common` and another for `lilac/overlay`.
`/device/sony/yoshino-common/hardware/OpenCustomizationSelector`
and the one you see in the file I changed.

Another question: where are the modem customization files located (e.g. `/etc/customization/modem/122/modem.conf`)?

**EDIT: Disregard the question, I think I found them.**

**Comment by antoniu200** [comment]
Can there be a universal modem customization, i.e. always use ir51_ir92, so adding each and every IMS provider is not needed?
Are there any known issues with using ir51_ir92 on some mobile provider?

**Comment by antoniu200** [comment]
Okay, my Lineage source is finally ready to build. I hope I won't get hit by huge issues. Fingers crossed.
I replaced multiple things, nothing significant, but I made a list of everything.

The only things left to do are for me to test and contact Flamefire to see if the AOSP base can be changed to 10.0.0_r47 instead of 10.0.0_r41, the latter of which I will do now.

**Comment by the-r3dacted** [comment]
> but OpenCustomizationSelector still resets the custom modem to default if the SIM is changed

I guess that should be the intended result for most people. Having it default to ir92 IMS though would probably be better overall for unrecognized SIMs.

> `/device/sony/yoshino-common/hardware/OpenCustomizationSelector`

I did not think to check in the hardware folder. I would like to port this, Xperia Parts, and whatever else to the XZ2C, but I'm stumbling through too many projects at the moment to investigate further. Maybe when I get tired enough of the slowness of the XA2 I'll try it idk.

> Are there any known issues with using ir51_ir92 on some mobile provider?

It's probably for the best you use the carrier provided mbn if possible, but for some carriers it seems the generic one works better idk how you mess up that bad. Some might have a highly modified implementation, like AT&T for example, but they have a whitelist so it doesn't really matter as you can't use it anyways.

**Comment by antoniu200** [comment]
I compiled the OS yesterday, tested today. DIGI now has VoLTE and VoWiFi working.
I, however, forgot to update `OpenCustomizationSelector/res/xml/configuration_selectors.xml`, which means the correct modem is still not applied automatically as of now. So, tomorrow, if I'll have any time, I will check my modifications again and compile and test yet again.

Overall, the changes are going in the right direction.
If only someone knew what that Navbar bug was all about...

**Comment by Flamefire** [comment]
> I compiled the OS yesterday, tested today. DIGI now has VoLTE and VoWiFi working. I, however, forgot to update `OpenCustomizationSelector/res/xml/configuration_selectors.xml`, which means the correct modem is still not applied automatically as of now. So, tomorrow, if I'll have any time, I will check my modifications again and compile and test yet again.

So you are saying that with the change from this PR it works? If so I'll include the change in the next builds by default.

> The only things left to do are for me to test and contact Flamefire to see if the AOSP base can be changed to 10.0.0_r47 instead of 10.0.0_r41, the latter of which I will do now.

I remember having tried that but stopped either due to too many issues or changes or due to LOS upstream sticking to r41 and the changes looked to big  given that 17.1 is pretty working (so I didn't want to break it)

**Comment by antoniu200** [comment]
> So you are saying that with the change from this PR it works? If so I'll include the change in the next builds by default.

Not yet. I have done more modifications offline. I will have to update the pull request once everything works as intended. The modem must switch automatically, otherwise it will be a pain to apply (3 reboots necessary when swapping SIM - explained above).

Currently compiling, probably one hour left. Will have to test, hopefully I can do that in the rest of today's time.

> I remember having tried that but stopped either due to too many issues or changes or due to LOS upstream sticking to r41 and the changes looked to big given that 17.1 is pretty working (so I didn't want to break it)

I have tried that 3 months ago as well, but did not have the time, nor the patience, to keep solving all the issues that kept coming up.
But it might be worth looking into compiling a stock AOSP and simply porting over the device tree, at least for testing purposes.

**Comment by antoniu200** [comment]
Okay, I have updated the PR. Take a look, especially at `overlay/vendor/lineage/bootanimation/CleanSpec.mk`, as I am not sure it's supposed to be there.
I copied the files straight from `/vendor/lineage/bootanimation` and it was there. Maybe it does nothing unless it's needed. But, for safety, please double-check my work.

Otherwise, I have:
1. Included [VoLTE Enabler](https://xdaforums.com/t/mod-magisk-root-volte-enabler.4372705/) in `system.prop`, since `ir51_ir92_ims_tar.mbn` does not work at all on DIGI Romania otherwise (worth looking further into why?);
2. Added (missing?) modem configurations, `930` and `931`, for Orange Romania and DIGI Romania, as neither of them had any (although Orange Romania had an MBN file 🤷, however only VoLTE works while using it);
3. Initially updated `XperiaParts/res/values/strings.xml` to correct a typo in `ims_summary`, but ended up changing that phrase altogether to reflect these (and upcoming?) IMS support changes;
4. Updated `XperiaParts/res/xml/device_settings.xml` to make the Customization Switcher notification default to off, instead of on, since its only use is debugging, which the general user does (and should do) very little of;
5. Last, but not least (you may choose to not add this), fixed and ported the Xperia X 720p bootanimation to the device tree.

**EDIT**: In conclusion, the ROM now has working VoLTE and VoWiFi for DIGI Romania and Orange Romania, with modem switching automatically, and, if you decide to include it, the 720p version of the full Xperia X bootanimation.
The bootanimation that came with the device from the factory was cut short and ran at a different than native resolution.

**Comment by the-r3dacted** [comment]
Why not default to ir51_ir92 so those modem configurations aren't needed at all?

**Comment by antoniu200** [comment]
> Why not default to ir51_ir92 so those modem configurations aren't needed at all?

That requires further digging. As of now, this works well.
Not to mention, `ir51_ir92` only enables IMS on mobile providers that keep their IMS structure very close to standard. It is not a universal solution, so some configurations will need to stay anyway.

I have not given up on that idea, but I don't have the time to dig and test that much right now (how does `ir51_ir92` fare on networks without IMS?). I am glad IMS works properly on these two networks, so far.

**Comment by the-r3dacted** [comment]
> Not to mention, `ir51_ir92` only enables IMS on mobile providers that keep their IMS structure very close to standard. It is not a universal solution, so some configurations will need to stay anyway.

Yeah I know, I specifically meant the two you added.

> how does `ir51_ir92` fare on networks without IMS?

I assume it would just work the same, but are there any non-IMS networks that exist?

**Comment by antoniu200** [comment]
> Yeah I know, I specifically meant the two you added.

These configs represent the way I tested IMS from the two providers on the G8441. They work nicely, so this is what I can offer **as of now**.
I will be looking into switching the default modem to `ir51_ir92`, but I don't know when my schedule will allow. Hence these configs.

> I assume it would just work the same, but are there any non-IMS networks that exist?

It actually does, indeed. I just limited the device to WCDMA and below with `ir51_ir92` enabled and it worked as expected.

As for networks without IMS, we don't know. But it's best to cover as many possibilities.

**Comment by antoniu200** [comment]
I think I identified the place where modifications need to happen for OpenCustomizationSelector to default to `ir51_ir92_ims`.

I am not sure, but my guess is that such important modification should be done [here](https://github.com/whatawurst/OpenCustomizationSelector), right at the source, along with some others that I submitted here.

I will test my assumption first, then modify this PR yet again accordingly:
- If I succeed in setting `ir51_ir92_ims` as default, I will open a PR over on OpenCustomizationSelector and retract/move `modem.conf`, `strings.xml` and some other related changes over there;
- If I do not, I will keep everything as is, **for now**.

Allow me 2-3 days of testing (10th of February).

**Comment by antoniu200** [comment]
Guys, do you own an Xperia XZ1 Compact or some other Xperia device IMS-compatible, but that did not have IMS services enabled on the network you are using currently?

I think I enabled all IMS services on (almost) all service providers, but I would need to test on more providers than Orange and DIGI Romania.
Could you guys help me?

EDIT: Maybe we could let the other users of this ROM know, and have some try a beta of the next release?

**Comment by the-r3dacted** [comment]
> Guys, do you own an Xperia XZ1 Compact or some other Xperia device IMS-compatible, but that did not have IMS services enabled on the network you are using currently?

Shit if you were a day sooner I would've said T-Mobile on Xperia XZ, but I don't know if anyone could get that working. Migrating to Verizon now. Wish my XZ1C didn't break because that would be fun to experiment with.

**Comment by the-r3dacted** [comment]
@antoniu200 How well do you know how IMS/VoLTE works in LineageOS? I'm having some trouble getting Verizon VoLTE working on the XZ2C. It seems to only work in 19.1+, and I'm trying to backport the change to 17.1/RR Q but I have no idea what the change is.

**Comment by antoniu200** [comment]
Check modem customizations, maybe one was added.
Check CarrierConfig, maybe a specific MCC / MNC one was added (to find MCC / MNC, view Customization Switcher logs).

Also, send me a link to your github workspace, we could talk there.

**Comment by antoniu200** [comment]
@Flamefire Sorry for tagging, but may I upload a .zip of a final, stable modified version that should support a whole bunch more mobile providers over on your XDA thread?
I would like to get some input from users and fix whatever bugs they may encounter.

**EDIT**: CustomizationSelector was also modified to default to `ir51_ir92_ims` when IMS Features are enabled and when no customizations are found. Link to my version: https://github.com/antoniu200/OpenCustomizationSelector/

**Comment by the-r3dacted** [comment]
> Check modem customizations, maybe one was added. Check CarrierConfig, maybe a specific MCC / MNC one was added (to find MCC / MNC, view Customization Switcher logs).
> 
> Also, send me a link to your github workspace, we could talk there.

Customization Switcher seems to be a 18.1+ thing. But T-Mobile VoLTE works fine without it, so I don't think that would be the issue but idk. I've confirmed MCC / MNC are correct in the APN though. I've also tried apns-conf.xml from 20.0 and that hasn't fixed it either. It's always just a Server unreachable error trying to place a call. Is there a way to get logs from 18.1 and 19.1 and see what went wrong and changed?

Also what is a GitHub Workspace?

**Comment by the-r3dacted** [comment]
@antoniu200 I will mention. My GitHub email (for now) just forwards to r3n@gmx.com since I don't have an email server set up. Gmail is not the best and may have marked the message as spam.

**Comment by the-r3dacted** [comment]
I wonder if reapplying the mbn at boot should be enabled by default. It's how it works atock afaik.

**Comment by antoniu200** [comment]
> @antoniu200 I will mention. My GitHub email (for now) just forwards to [r3n@gmx.com](mailto:r3n@gmx.com) since I don't have an email server set up. Gmail is not the best and may have marked the message as spam.

Oh, wow, I only just now saw your email. I thought you meant in this message that you have **no** email address, don't know how I understood that. I'll reply to you later tonight.

> I wonder if reapplying the mbn at boot should be enabled by default. It's how it works atock afaik.

No idea how it works on stock, haven't used it since Dec. 2021. All I know is that `ir51_ir92` sometimes crashes after airplane mode and reapplying the modem fixes that (it's reapplied both at boot and after airplane mode).

**Comment by Flamefire** [comment]
> @Flamefire Sorry for tagging, but may I upload a .zip of a final, stable modified version that should support a whole bunch more mobile providers over on your XDA thread? I would like to get some input from users and fix whatever bugs they may encounter.

Feel free. However I don't think removing the carrier specific configuration as you did here is the right way. It might create problems where it was working before, especially as some carriers seem to need non-default `carrier_default_wfc_ims_mode_int`

So I'd rather add to the list as required than blankly replacing everything and hoping for the best.   
Also this PR contains unrelated changes to `overlay/device/sony/yoshino-common/hardware/OpenCustomizationSelector`, `overlay/device/sony/yoshino-common/hardware/XperiaParts` and `lineage_lilac.mk` which don't look to be intended and are superflous as far as I can tell.

> **EDIT**: CustomizationSelector was also modified to default to `ir51_ir92_ims` when IMS Features are enabled and when no customizations are found. Link to my version: https://github.com/antoniu200/OpenCustomizationSelector/

Looks like you've been using the wrong starting point.
The currently used version is at https://github.com/Flamefire/android_device_sony_yoshino-common/tree/lineage-17.1/hardware/OpenCustomizationSelector originally forked from https://github.com/whatawurst/android_device_sony_yoshino-common which contains a couple more changes. IIRC the repo you forked was mostly a decompiled APK or something like that as the initial base. Also your large commits make it hard to see what you actually intended, e.g. https://github.com/antoniu200/OpenCustomizationSelector/commit/a50b64d6844709bd5c07f6342938a0dbe2450348 So I don't know what actually needs changing.

**Comment by antoniu200** [comment]
> Feel free. However I don't think removing the carrier specific configuration as you did here is the right way. It might create problems where it was working before, especially as some carriers seem to need non-default `carrier_default_wfc_ims_mode_int`

Looking at documentation, that is actually just the default way Wi-Fi calling should be used: only when LTE is unavailable or always when Wi-Fi is available. I don't believe carriers can check Android settings, and, even if they can, I doubt they *need* it configured like that, since users may change that in Settings.

> So I'd rather add to the list as required than blankly replacing everything and hoping for the best.

There's no way we can index all mobile carriers worldwide (basically, what adding like Sony's done would imply), so, if we want to get as many carriers working, we need to make one universal carrier config, and, maybe, add on top of that where corrections are needed.

IMS registration is done by modem (OpenCustomizationSelector). CarrierConfig simply tells Android what the carrier (and phone) supports and / or needs (provisioning).

Other options than what I enabled simply set default Settings - they're useless to registering IMS.

> Also this PR contains unrelated changes to `overlay/device/sony/yoshino-common/hardware/OpenCustomizationSelector`, `overlay/device/sony/yoshino-common/hardware/XperiaParts` and `lineage_lilac.mk` which don't look to be intended and are superflous as far as I can tell.

I'll have to take a look. IIRC, the OpenCustomizationSelector modification here contains a fixed translation, while XperiaParts needs some text changed for an option (IMS would now work on more carriers than in stock) and Reapply Modem now needs to default to enabled, since `ir51_ir92` crashes when rebooting and toggling airplane mode.

I'll remove `lineage_lilac.mk`, that was modified to accomodate the Xperia bootanimation.

> Looks like you've been using the wrong starting point. The currently used version is at https://github.com/Flamefire/android_device_sony_yoshino-common/tree/lineage-17.1/hardware/OpenCustomizationSelector originally forked from https://github.com/whatawurst/android_device_sony_yoshino-common which contains a couple more changes. IIRC the repo you forked was mostly a decompiled APK or something like that as the initial base. Also your large commits make it hard to see what you actually intended, e.g. [antoniu200/OpenCustomizationSelector@a50b64d](https://github.com/antoniu200/OpenCustomizationSelector/commit/a50b64d6844709bd5c07f6342938a0dbe2450348) So I don't know what actually needs changing.

Oh, sorry, good to know. I'll redo the OpenCustomizationSelector commits then.

I also changed `RadioInfo.java` and `ImsPhoneCallTracker.java`.
I will upload those too later.

**Comment by Flamefire** [comment]
> Looking at documentation, that is actually just the default way Wi-Fi calling should be used: only when LTE is unavailable or always when Wi-Fi is available. I don't believe carriers can check Android settings, and, even if they can, I doubt they _need_ it configured like that, since users may change that in Settings.

Ah true.

> There's no way we can index all mobile carriers worldwide (basically, what adding like Sony's done would imply), so, if we want to get as many carriers working, we need to make one universal carrier config, and, maybe, add on top of that where corrections are needed.

Ok so I'd guess a safe(r) approach would be to add those defaults you added (<carrier_config>https://github.com/Flamefire/android_device_sony_lilac/pull/23/files#diff-d7fb4c78ce6da621e92b4ff90608a8e2ddcd87fb3a83edc32f08ea5dca9430a2R72-R88) and remove them from the individual carriers. E.g. remove all `<boolean name="carrier_volte_available_bool" value="true" />` lines. Then remove all (now) empty carrier configs. So we keep stuff like `apn_hide_rule_strings_with_iccids_array`

> IMS registration is done by modem (OpenCustomizationSelector). CarrierConfig simply tells Android what the carrier (and phone) supports and / or needs (provisioning).

I just fear that enabling stuff that isn't supported will then later break on usage in very unexpected ways if Android relies on that information being correct.

> I'll have to take a look. IIRC, the OpenCustomizationSelector modification here contains a fixed translation, while XperiaParts needs some text changed for an option (IMS would now work on more carriers than in stock) and Reapply Modem now needs to default to enabled, since `ir51_ir92` crashes when rebooting and toggling airplane mode.

I added the [fixed translation](https://github.com/Flamefire/android_device_sony_yoshino-common/commit/ffd80d30148418bd5e12c082a6e7eaea91ce66d3) to my branch. The original one reads better through google translate but I trust your judgement more here ;-)

> Oh, sorry, good to know. I'll redo the OpenCustomizationSelector commits then.
> 
> I also changed `RadioInfo.java` and `ImsPhoneCallTracker.java`. I will upload those too later.

Ok. Please try to make the commits as small (but still self-contained) as possible such that one can see what/why is done and revert individual changes when required.

**Comment by antoniu200** [comment]
> Ok so I'd guess a safe(r) approach would be to add those defaults you added (<carrier_config>https://github.com/Flamefire/android_device_sony_lilac/pull/23/files#diff-d7fb4c78ce6da621e92b4ff90608a8e2ddcd87fb3a83edc32f08ea5dca9430a2R72-R88) and remove them from the individual carriers. E.g. remove all `<boolean name="carrier_volte_available_bool" value="true" />` lines. Then remove all (now) empty carrier configs. So we keep stuff like `apn_hide_rule_strings_with_iccids_array`

I just eliminated networks that were covered by the universal config, and went through the remaining ones to see if their strings are documented in `CarrierConfigManager.java`. They are not, which means they no longer work anyway, or maybe they never did and Sony were deploying a modified CarrierConfig app.

So, they are pointless anyway.

> I just fear that enabling stuff that isn't supported will then later break on usage in very unexpected ways if Android relies on that information being correct.

Android does rely on that information being correct, and it is: the phone supports all technologies listed in `vendor.xml`. As for registration with mobile carriers, it is done by provisioning. So, only technologies that are supported by the mobile carriers will end up being used and enabled.

For carriers that do not offer provisioning servers, I modified `ImsPhoneCallTracker.java`, so only technologies where the phone is registered by the carrier will be enabled and set as provisioned (uploading modified code and CustomizationSelector commits right now).

**Comment by antoniu200** [comment]
OpenCustomizationSelector is done: https://github.com/antoniu200/android_device_sony_yoshino-common/tree/lineage-18.1/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector
Also moved other files that belonged there, opened a pull request over at `android_device_sony_yoshino-common`.

However, I am not sure that `lineage-18.1` is the branch downloaded when doing `repo sync` on 17.1, as I have *local* files in OpenCustomizationSelector that look like they're from `checked` branch, and others from `lineage-18.1`. Please check on your end as well.

**Comment by antoniu200** [comment]
https://github.com/antoniu200/android_packages_apps_Settings/blob/lineage-17.1/src/com/android/settings/RadioInfo.java
https://github.com/antoniu200/android_frameworks_opt_telephony/blob/lineage-17.1/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java

It's all done. Unfortunately, I do not know how to make a patch file, so I hope the above links will do.

**Comment by the-r3dacted** [comment]
Requires a few tweaks to work with applyPatches.sh, but GitHub can do it for you.

https://github.com/antoniu200/android_packages_apps_Settings/commit/907d5897ed90bf1e210c6c42bd958eac388e6c5d.patch

https://github.com/antoniu200/android_frameworks_opt_telephony/commit/959dcadfec2f7a44f8d884c28f9fa5321ddf6ee0.patch

**Comment by Flamefire** [comment]
> I just eliminated networks that were covered by the universal config, and went through the remaining ones to see if their strings are documented in CarrierConfigManager.java. They are not, which means they no longer work anyway, or maybe they never did and Sony were deploying a modified CarrierConfig app.

Note that CarrierConfigManager.java is not the only one checking that. There is e.g. vendor/codeaurora/telephony/ims/src/org/codeaurora/ims/QtiCarrierConfigs.java which defines `video_call_use_ext` or `use_custom_video_ui`

> OpenCustomizationSelector is done: https://github.com/antoniu200/android_device_sony_yoshino-common/tree/lineage-18.1/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector Also moved other files that belonged there, opened a pull request over at `android_device_sony_yoshino-common`.

I can't find the pull request. Can you post a link to it? And do you intent the changes for 17.1 or 18.1? Because your link is in the 18.1 branch, not 17.1

> However, I am not sure that `lineage-18.1` is the branch downloaded when doing `repo sync` on 17.1, as I have _local_ files in OpenCustomizationSelector that look like they're from `checked` branch, and others from `lineage-18.1`. Please check on your end as well.

I'm not sure how you ended up with a totally unrelated branch there. If you are on 17.1 then your manifest should contain `<project name="Flamefire/android_device_sony_yoshino-common" path="device/sony/yoshino-common" remote="github" revision="lineage-17.1" />` What do you have in there?

> https://github.com/antoniu200/android_packages_apps_Settings/blob/lineage-17.1/src/com/android/settings/RadioInfo.java

This seems to disable setting switches for provisioning. However you can only access that setting screen via a "secret code" and they are already disabled in LOS builds, aren't they? So why bother with that change which could cause conflicts and hence more work?

> https://github.com/antoniu200/android_frameworks_opt_telephony/blob/lineage-17.1/src/java/com/android/internal/telephony/imsphone/ImsPhoneCallTracker.java

Looks OK I guess. Although I'd adjust the conditions a bit to take advantage of short-circuiting and possibly save a bit of battery life in case that code ends up being called frequently.

File: overlay/packages/apps/CarrierConfig/res/xml/vendor.xml [hunk 1]
```diff
@@ -60,1542 +60,32 @@ IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.-->
...
-        <boolean name="carrier_volte_available_bool" value="true" />
+    
+    <carrier_config>
+	<!-- Make VoLTE, VoWiFi and Video Telephony "available" or "enabled by platform". -->	
+	<boolean name="carrier_volte_available_bool" value="true" />
+	<boolean name="carrier_wfc_ims_available_bool" value="true" />
+	<boolean name="carrier_vt_available_bool" value="true" />
+	<boolean name="use_rcs_presence_bool" value="true" />
+	
+	<!-- Enable provisioning, so we do not reach 'the end of the world' by unavailable features. -->
```
Comment (Flamefire) [inline overlay/packages/apps/CarrierConfig/res/xml/vendor.xml]: What is meant be "reaching the end of the world" here and below?

Comment (antoniu200) [reply overlay/packages/apps/CarrierConfig/res/xml/vendor.xml]: Downgrade techs if unavailable, or don't provision unavailable techs. Otherwise, user gets busy tone (end of the world - a joke if you will).

**Comment by antoniu200** [comment]
> Note that CarrierConfigManager.java is not the only one checking that. There is e.g. vendor/codeaurora/telephony/ims/src/org/codeaurora/ims/QtiCarrierConfigs.java which defines `video_call_use_ext` or `use_custom_video_ui`

Ok, I'll check again.

> I can't find the pull request. Can you post a link to it? And do you intent the changes for 17.1 or 18.1? Because your link is in the 18.1 branch, not 17.1

https://github.com/Flamefire/android_device_sony_yoshino-common/pull/38

I made modifications in 18.1, as that is the default branch, and it's similar with what I have offline. I have to check.

> This seems to disable setting switches for provisioning. However you can only access that setting screen via a "secret code" and they are already disabled in LOS builds, aren't they? So why bother with that change which could cause conflicts and hence more work?

`*#*#4636#*#*` is not disabled and is useful. However, the switches are enabled in non-user builds, which means users can break IMS with them.

I see no other conflicts.

> Looks OK I guess. Although I'd adjust the conditions a bit to take advantage of short-circuiting and possibly save a bit of battery life in case that code ends up being called frequently.

That code I modified is called when switching from Mobile Data to Wi-Fi and back.

Not sure what other optimizations I could do.

**Comment by antoniu200** [comment]
I opened pull requests for the other versions as well.
Why do we need 3 different, but very similar, versions of the `yoshino-common` device tree? You were saying a while back that you're using the 17.1 device tree for all versions.

https://github.com/Flamefire/android_device_sony_yoshino-common/pull/39
https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40

> I'm not sure how you ended up with a totally unrelated branch there. If you are on 17.1 then your manifest should contain `<project name="Flamefire/android_device_sony_yoshino-common" path="device/sony/yoshino-common" remote="github" revision="lineage-17.1" />` What do you have in there?

```
[remote "github"]
	url = https://github.com/Flamefire/android_device_sony_yoshino-common
	review = review.lineageos.org
	projectname = Flamefire/android_device_sony_yoshino-common
	fetch = +refs/heads/*:refs/remotes/github/*
```

This is all I found, it's in `/device/sony/yoshino-common/.git/config`. Not sure where else I should look. `/.repo/manifest.xml` has no repositories and `/.repo/manifests/default.xml` contains no `Flamefire/android_device_sony_yoshino-common`.

**Comment by Flamefire** [comment]
> I opened pull requests for the other versions as well. Why do we need 3 different, but very similar, versions of the `yoshino-common` device tree? You were saying a while back that you're using the 17.1 device tree for all versions.

You only need one PR. We can discuss the changes there and I'll cherry-pick to the other branches when done.

You might have misunderstood. The device trees have to be different although only slightly due to different requirements by Android. I remember saying that I "build **mostly** from the same sources" or something like that. I guess that's what you meant.

> This is all I found, it's in `/device/sony/yoshino-common/.git/config`. Not sure where else I should look. `/.repo/manifest.xml` has no repositories and `/.repo/manifests/default.xml` contains no `Flamefire/android_device_sony_yoshino-common`.

You must have added that entry otherwise `repo sync` wouldn't sync that folder would it? You can use `grep` to search for the manifest where you added it, but I suppose it is in /.repo/local_manifests/roomservice.xml

**Comment by the-r3dacted** [comment]
Need to test on XZ1C once I get it repaired, but I got AT&T VoLTE working on XA2 with ir51_ir92 mbn.

More detail here. https://github.com/Eclipse-Community/build_lineage/issues/9

**Comment by the-r3dacted** [comment]
Not entirely sure why, but I can hack AT&T VoLTE onto XZ1C stock and have it work, but as soon as I custom ROM I just get no signal at all.

Trying to manually pick network just hangs on Connecting. Sometimes I get a No service Selected network unavailable notification but the choose network screen is still up and acting like it's connecting. If I back out and go back I get Couldn't find networks. Try again.

**Comment by the-r3dacted** [comment]
I can't just swap SIMs on lilac even though OpenCustomizationSelector should just let that happen. I always just get no signal. Even from same carrier seemingly. I had to reflash stock, verify it works with T-Mobile, then reinstall custom ROM, and now I have T-Mobile signal. I wasn't able to go from AT&T to T-Mobile directly.

**Comment by antoniu200** [comment]
I am noticing strange things on my end as well, with IMS services not working anymore on newly added carriers, for whatever reason.
I was looking into the issue, but my XZ1c's USB port died, so I cannot test anymore.

Hoping to get the phone fixed next week. Maybe try to force enhanced 4G LTE to be enabled meanwhile?

**Comment by the-r3dacted** [comment]
Could try cleaning the port. If you can get power working, then you can still sideload with MicroSD.

How do I force enhanced 4G LTE? Also what is enhanced 4G LTE?

**Comment by the-r3dacted** [comment]
So I noticed T-Mobile SIM was connecting only in GSM until forced to be LTE? I have the 2G setting off and LTE preferred set. Not sure why it's doing this.

**Comment by the-r3dacted** [comment]
Wonder if related to modem bug. Set at boot to switch to 2G, and it near instantly got 2G before switching to 4G.

What causes the modem bug? Why is it seemingly not an issue stock?

**Comment by antoniu200** [comment]
Okay, I am back, finally fixed my phone. The port had come semi-unplugged from the mainboard, probably because I had connected it poorly to begin with.

Back to the topic: Enhanced 4G LTE turns out to be...
You won't believe it: **VoLTE**. It was not the cause of my bug.

My bug was caused by the fact that I had assumed IMS registration is done fully by the modem, as long as IMS calling options are "enabled by platform" (i.e. enabled in this `vendor.xml`). I was wrong, the user must also enable them in Network Settings.

But how does the user know whether or not they should enable a particular IMS calling setting? It is assumed the system will only show available techs, so that's how an average user knows.

Hence, we cannot rely on the user. **Solution**: set VoLTE and VoWiFi as enabled by default in Settings (will not cause toggles to appear), and simply refuse their provisioning, if required.

Basically, add 2 more lines to this `vendor.xml`.

**Comment by the-r3dacted** [comment]
Awesome. However I don't want to test at least until I get a spare XZ1C. I've gone back to maining the XZ1C and don't want to mess the current config up.

**Comment by the-r3dacted** [comment]
Wow. I'll give it a bit to be tested, then compile my RR Q and try it out.

**Comment by the-r3dacted** [comment]
I took a bit of a leap of faith and just tried the FreedomPop SIM in my XZ1C with my updated compile, and it just worked. I did have to set it from 3G to LTE manually, but it just worked, and I was able to swap back to T-Mobile and it works fine still. Amazing. If I find a better provider that also ignores the whitelist, I will switch.
